pincer/commands/chat_command_handler.py

```diff
@@ -0,0 +1,435 @@
+                    type=AppCommandOptionType.SUB_COMMAND_GROUP,
+                    options=[]
+                )
+
+                # This for-else makes sure that sub_command_group will hold a reference
+                # to the subcommand group that we want to modify to hold `cmd`
+
+                for cmd_in_children in children:
+                    if (
+                        cmd_in_children.name == sub_command_group.name
+                        and cmd_in_children.description == sub_command_group.description
+                        and cmd_in_children.type == sub_command_group.type
+                    ):
+                        sub_command_group = cmd_in_children
+                        break
+                else:
+                    children.append(sub_command_group)
+
+                sub_command_group.options.append(AppCommandOption(
+                    name=cmd.app.name,
+                    description=cmd.app.description,
+                    type=AppCommandOptionType.SUB_COMMAND,
+                    options=cmd.app.options,
+                ))
+
+                continue
+
+            if cmd.group:
+                # Any command at this point will only have one level of nesting.
+                #
+                # Command
+                #    subcommand
+                #
+                # A subcommand object is what is being generated here. If there is no
+                # top level command, it will be created here.
+
+                # `key` represents the hash value for the top-level command that will
+                # hold the subcommand.
+
+                key = _hash_app_command_params(
+                    cmd.group.name,
+                    cmd.app.guild_id,
+                    AppCommandOptionType.SUB_COMMAND,
+                    None,
+                    None
+                )
+
+                if key not in ChatCommandHandler.built_register:
+                    ChatCommandHandler.built_register[key] = AppCommand(
+                        name=cmd.group.name,
+                        description=cmd.group.description,
+                        type=AppCommandOptionType.SUB_COMMAND,
+                        guild_id=cmd.app.guild_id,
+                        options=[]
+                    )
+
+                # No checking has to be done before appending `cmd` since it is the
+                # lowest level.
+                ChatCommandHandler.built_register[key].options.append(
+                    AppCommandOption(
+                        name=cmd.app.name,
+                        description=cmd.app.description,
+                        type=AppCommandType.CHAT_INPUT,
+                        options=cmd.app.options
+                    )
+                )
+
+                continue
+
+            # All single-level commands are registered here.
+            ChatCommandHandler.built_register[
+                _hash_app_command(cmd.app, cmd.group, cmd.sub_group)
```

Comment:
Why is _hash_app_command not part of the cmd.app?
Or like
```py
hash(cmd.app)
```
?

Reply:
the function is "translated" to 
```py
hash(cmd.app.name, cmd.app.guild_id, cmd.app.type, cmd.group, cmd.sub_group)
```

Reply:
```py
class Foo:

    def __init__(self):
        self.a = 1
        self.b = 2

    def __hash__(self):
        return hash((self.a, self.b))


print(hash(Foo()))
```
This does works